Mark Obsoletions for .NET 5

jeffhandley opened the pull request:
This contributes the remaining obsoletions from https://github.com/dotnet/designs/pull/139.

* `SYSLIB0003` - Code Access Security is not supported or honored by the runtime.
* `SYSLIB0004` - The Constrained Execution Region (CER) feature is not supported.
* `SYSLIB0005` - The Global Assembly Cache is not supported.
* `SYSLIB0006` - Thread.Abort is not supported and throws PlatformNotSupportedException.
* `SYSLIB0007` - The default implementation of this cryptography algorithm is not supported.
* `SYSLIB0008` - The CreatePdbGenerator API is not supported and throws PlatformNotSupportedException.
* `SYSLIB0009` - The AuthenticationManager Authenticate and PreAuthenticate methods are not supported and throw PlatformNotSupportedException.
* `SYSLIB0010` - This Remoting API is not supported and throws PlatformNotSupportedException.

Existing obsoletions were updated to reflect the diagnostic id prefix of `SYSLIB`. `BinaryFormatter` was moved to `SYSLIB0011`.

jeffhandley committed:
Mark new obsoletions for .NET 5; update existing obsoletions

stephentoub commented on docs/project/list-of-obsoletions.md:
```diff
@@ -15,3 +15,5 @@ Currently the identifiers `MSLIB0001` through `MSLIB0999` are carved out for obs
 |  __`MSLIB0001`__ | The UTF-7 encoding is insecure and should not be used. Consider using UTF-8 instead. |
 |  __`MSLIB0002`__ | `PrincipalPermissionAttribute` is not honored by the runtime and must not be used. |
+|  __`MSLIB0004`__ | The Constrained Execution Region (CER) feature is no longer implemented. |
+|  __`MSLIB0005`__ | The Global Assembly Cache is no longer implemented |
```
```suggestion
|  __`MSLIB0005`__ | The Global Assembly Cache is no longer implemented. |
```

stephentoub commented on src/libraries/Common/tests/TestUtilities/System/Buffers/BoundedMemory.Windows.cs:
```diff
@@ -305,7 +305,9 @@ private static class UnsafeNativeMethods
             [DllImport(KERNEL32_LIB, CallingConvention = CallingConvention.Winapi, SetLastError = true)]
+#if FEATURE_CER
             [ReliabilityContract(Consistency.WillNotCorruptState, Cer.Success)]
+#endif
```
Why do we need this attribute at all?

jeffhandley replied:
I wasn't positive whether or not there was a build that would utilize it; but I'll remove it and then I can remove the constant from the csproj too.

jkotas commented on src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/EventLog.cs:
```diff
@@ -265,7 +265,9 @@ public static void CreateEventSource(EventSourceCreationData sourceData)
                 throw new ArgumentException(SR.Format(SR.ParameterTooLong, nameof(source), 254 - EventLogKey.Length));
 
             Mutex mutex = null;
+#if FEATURE_CER
```
This file is compiled for .NET Core only. We do not need this under ifdefs. This should be deleted instead.

(multiple places)

jeffhandley replied:
Ah, that makes sense. I was leaving the calls in place for `netcoreapp2.0`, but since it's a no-op it makes sense that it can be removed.  Thanks!

jeffhandley replied:
And on that note, I think I can change my pattern for how I define `FEATURE_CER` such that it's only defined when `netstandard` or `net4` are targeted.

jkotas commented on src/libraries/System.Data.Odbc/src/Common/System/Data/Common/AdapterUtil.Odbc.cs:
```diff
@@ -656,7 +656,9 @@ internal static int StringLength(string inputString)
             return ((null != inputString) ? inputString.Length : 0);
         }
 
+#if FEATURE_CER
```
This file compiles for netcoreapp only. You can tell from this condition in the .csproj:
```
  <ItemGroup Condition="($(TargetFramework.StartsWith('$(NetCoreAppCurrent)')) or $(TargetFramework.StartsWith('netcoreapp2.0'))) and '$(TargetsAnyOS)' != 'true'">
```
-> this can be deleted.

jkotas replied:
(multiple places)

GrabYourPitchforks commented on src/libraries/Directory.Build.targets:
```diff
@@ -8,6 +8,7 @@
   <PropertyGroup>
     <NoWarn Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework' or '$(TargetFrameworkIdentifier)' == '.NETStandard' or '$(TargetFramework)' == 'netcoreapp2.1'">$(NoWarn);nullable</NoWarn>
     <NoWarn Condition="'$(GeneratePlatformNotSupportedAssembly)' == 'true' or '$(GeneratePlatformNotSupportedAssemblyMessage)' != ''">$(NoWarn);nullable</NoWarn>
+    <NoWarn Condition="'$(IsPartialFacadeAssembly)' == 'true'">$(NoWarn);MSLIB0004</NoWarn>
```
I'd put a `<!-- comment -->` just above this line clarifying what _MSLIB0004_ is.

GrabYourPitchforks replied:
I'd put a `<!-- comment -->` just above this line clarifying what _MSLIB0004_ is.

GrabYourPitchforks commented on src/libraries/Microsoft.Win32.SystemEvents/src/Microsoft.Win32.SystemEvents.csproj:
```diff
@@ -1,4 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Double-check that a UTF-8 BOM didn't get added to this file.

jeffhandley replied:
Good catch; thanks!

GrabYourPitchforks commented on src/libraries/Microsoft.Win32.SystemEvents/src/Microsoft.Win32.SystemEvents.csproj:
```diff
@@ -8,6 +8,7 @@
   </PropertyGroup>
   <!-- DesignTimeBuild requires all the TargetFramework Derived Properties to not be present in the first property group. -->
   <PropertyGroup>
+    <DefineConstants Condition="$(TargetFramework.StartsWith('netstandard2.0')) or $(TargetFramework.StartsWith('netcoreapp2.0')) or $(TargetFramework.StartsWith('netcoreapp3.0'))">$(DefineConstants);FEATURE_CER</DefineConstants>
```
I'd actually turn this condition on its head. If the target framework starts with anything _other_ than `"netcore"`, define the constant. These APIs are no-ops across all versions of .NET Core, so excluding netcoreapp2.0/3.0 from containing the calls is ok.

It also helps make this a bit more robust, as we don't have to worry about "wait, does this project also target NetFramework?" This project does, and Full Framework should get CER calls.

(This comment applies to other .csproj files in this PR as well.)

jeffhandley replied:
I considered that but chose to go with an approach that uses a fixed list of known, past targets so that future versions are excluded by default.

GrabYourPitchforks commented on src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj:
```diff
@@ -1,4 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
UTF-8 BOM added by accident?

GrabYourPitchforks commented on src/libraries/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs:
```diff
@@ -301,10 +301,14 @@ internal static bool GenerateSerializerToStream(XmlMapping[] xmlMappings, Type[]
                             continue;
                         }
 
+#if FEATURE_GAC
```
Recommend instead using a single `#if` block instead of two blocks:

```cs
#if FOO
    if (condition)
#endif
    {
        /* logic */
    }
```

jeffhandley commented on src/libraries/System.Data.Odbc/src/System.Data.Odbc.csproj:
```diff
@@ -13,46 +13,32 @@
     <NoWarn Condition="'$(TargetFramework)' == '$(NetCoreAppCurrent)' and '$(TargetsAnyOS)' == 'true'">$(NoWarn);SA1121</NoWarn>
```
Yikes -- will fix

jkotas commented on src/libraries/Microsoft.Win32.SystemEvents/src/Microsoft.Win32.SystemEvents.csproj:
```diff
@@ -8,6 +8,7 @@
   </PropertyGroup>
   <!-- DesignTimeBuild requires all the TargetFramework Derived Properties to not be present in the first property group. -->
   <PropertyGroup>
+    <DefineConstants Condition="$(TargetFramework.StartsWith('netstandard2.0')) or $(TargetFramework.StartsWith('netcoreapp2.0')) or $(TargetFramework.StartsWith('netcoreapp3.0'))">$(DefineConstants);FEATURE_CER</DefineConstants>
```
The CER APIs were no-ops in netcoreapp2.0 and necoreapp3.0 as well. It is not necessary to define `FEATURE_CER` for those.

jeffhandley replied:
Thanks -- removing it everywhere.  Also seeing that FEATURE_GAC was not needed either.

jeffhandley replied:
It looks like the only place `FEATURE_CER` is appropriate is in `System.Reflection.Metadata`, where it targets `netstandard1.1` and `netstandard2.0` and compiles the source files for those targets.

jeffhandley replied:
And same is true for `FEATURE_GAC` -- `System.Reflection.Metadata` looks like the only necessary place for it.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Obsoletions.cs:
```diff
@@ -9,5 +9,14 @@ internal static class Obsoletions
+        internal const string GlobalAssemblyCacheMessage = "The Global Assembly Cache is no longer implemented.";
+        internal const string GlobalAssemblyCacheDiagId = "MSLIB0005";
+
+        internal const string ThreadAbortMessage = "Thread.Abort is no longer supported and throws PlatformNotSupportedException.";
```
This says "no longer supported" and the other ones say "no longer implemented". Should we say "no longer supported" everywhere?

The docs usually say "supported". E.g. https://docs.microsoft.com/en-us/dotnet/framework/performance/constrained-execution-regions

jkotas commented on src/libraries/System.DirectoryServices/src/System/DirectoryServices/DirectoryServicesPermission.cs:
```diff
@@ -5,6 +5,9 @@
 
 namespace System.DirectoryServices
 {
+#if CAS_OBSOLETIONS
```
Why can't this be applied unconditionally? 

If this needs to be this way, maybe comment explaining why this is applied in src under ifdef, but missing in the reference assembly may be useful. It is not obvious to me.

jeffhandley replied:
I saw that we had `_$(NetFrameworkCurrent)` as a target framework and I wasn't positive of the behavior of that. But talking with @safern last night, I understand it better. We are not actually compiling against .NET Framework there; this is for packaging purposes.  So you're right--I'll make it unconditional.  Thanks!

jeffhandley replied:
@safern helped me understand the `ref` assembly aspect of `System.DirectoryServices` here too.  We need to start targeting `$(NetCoreAppCurrent)` in that ref assembly, which we didn't do before because the surface area had not yet diverged from `netstandard2.0`.  I'm going to address that as well.

jeffhandley replied:
I can't get this working (today).  I'll need some infrastructure help getting `System.DirectoryServices` diverged so that its `ref` assembly can get built against `$(NetCoreAppCurrent)`.  Trying to do that led to transitive dependency challenges.

@GrabYourPitchforks and @eerhardt recommended we omit the DirectoryServices obsoletions from this PR and file an issue for RC1 to get the infrastructure updated to support adding them back.  I'm going to update this PR to reflect that and once that is looking solid, I'll submit an issue.

jeffhandley replied:
I filed https://github.com/dotnet/runtime/issues/39401 for the infrastructure need. Assuming this PR gets merged before that is fixed, I'll file another tracking issue for marking the DirectoryServices types as obsolete.

jeffhandley replied:
#39413 filed to mark those types as obsolete

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/TraceLoggingEventSource.cs:
```diff
@@ -440,9 +440,6 @@ public unsafe void Write(string? eventName, EventSourceOptions options)
                 descriptors[1].SetMetadata(pMetadata1, nameInfo.nameMetadata.Length, 1);
                 descriptors[2].SetMetadata(pMetadata2, eventTypes.typeMetadata.Length, 1);
 
-#if (!ES_BUILD_PCL)
```
This file is built as OOB package for .NET Framework. (`src\libraries\Microsoft.Diagnostics.Tracing.EventSource.Redist\src\Microsoft.Diagnostics.Tracing.EventSource.Redist.csproj`)

I think this should stay, perhaps under a different ifdef (`#if ES_BUILD_STANDALONE` ?). Make sure to fix all places in this file.

jeffhandley replied:
I connected with @noahfalk on this. We can't find the definition of `ES_BUILD_PCL` but we see it's referenced in some other places too.  We agreed that for this PR, I should:

1. Change this to `#if (!ES_BUILD_PCL && !NETCOREAPP)`
1. Submit an issue to the `area-System.Diagnostics.Tracing` area to investigate the potentially stale `ES_BUILD_PCL` references

jeffhandley replied:
I filed https://github.com/dotnet/runtime/issues/39412 to investigate the potentially unused constant.

jkotas commented on src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/Assemblies/RoAssembly.cs:
```diff
@@ -145,6 +145,9 @@ public sealed override AssemblyName[] GetReferencedAssemblies()
 
         // Miscellaneous properties
         public sealed override bool ReflectionOnly => true;
+#if !FEATURE_GAC
```
Can this be applied unconditionally? (Same as my other comment.)

jeffhandley replied:
I think this one needs to remain conditional because we build this for `netcoreapp3.0` and `netstandard2.0` too.

@steveharter - Can you confirm that I need this `#ifdef` to apply `[Obsolete]` to `RoAssembly.GlobalAssemblyCache` only in `net5.0`+ only?

steveharter replied:
I see no reason to have the condition. MetadataLoadContext doesn't provide a way to "resolve" an assembly from the GAC (even if running under .NET Framework) so this will always be `false`.

jeffhandley replied:
Thanks. I ended up finding it does still need to be conditional, since the DiagnosticId and UrlFormat properties are new in net5.0 and this compiles against lower versions.  I'm going to change the logic to be `#if GAC_OBSOLETIONS` though, so it connotes the reasoning more clearly.

jkotas commented on src/libraries/System.Security.Permissions/ref/System.Security.Permissions.Forwards.cs:
```diff
@@ -17,3 +18,5 @@
 [assembly: System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Security.Permissions.PermissionState))]
 #endif
+
+#pragma warning restore SYSLIB0003
```
Nit: I would not bother with restore at the end in these Forwards files (more than one place).

jeffhandley commented:
It looks like there are at least a few legitimate test failures from this where compensating changes need to be made for the obsoleted APIs. I'll look at those later this morning.

jkotas commented on src/libraries/System.Security.Permissions/src/System/Security/Permissions/Obsoletions.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    internal static class Obsoletions
+    {
+        internal const string SharedUrlFormat = "https://aka.ms/dotnet-warnings/{0}";
```
Should be have the Obsoletions.cs file under Common to avoid duplication?

(The IL linker should trim the unused strings literals, so including all strings into each assembly with obsoletions should not be a problem.)

jeffhandley committed:
Exclude DirectoryServicesPermission/DirectoryServicesPermissionAttribute from obsoletions for now

jeffhandley committed:
No need to restore SYSLIB0003 at end of Forwards files

jeffhandley committed:
Use a common Obsoletions.cs file for all SYSLIB obsoletions

jeffhandley committed:
Resurrect CER code behind #ifdef in EventSource

jeffhandley committed:
Mark GlobalAssemblyCache overrides as Obsolete

jeffhandley committed:
Suppress Thread.Abort() error in ImageAnimator.Unix.cs

jeffhandley committed:
Unconditionally mark RoAssembly.GlobalAssemblyCache as obsolete

jeffhandley committed:
Rename MSLIB0001 to SYSLIB0001 in ref assembly

jeffhandley committed:
Suppress obsoletion warnings in tests

jeffhandley committed:
Adopt a uniform #ifdef for NET50_OBSOLETIONS

jeffhandley committed:
Fix BinaryFormatter obsoletion message after the rebase

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Security/IStackWalk.cs:
```diff
@@ -3,6 +3,9 @@
 
 namespace System.Security
 {
+#if NETCOREAPP // This file is compiled into both System.Private.CoreLib (NetCoreApp) and System.Security.Permission (netstandard2.0)
```
```suggestion
#if SYSTEM_PRIVATE_CORELIB
```
It will make the code self-describing. We are using this ifdef in number of other similar situations in CoreLib.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Security/PermissionSet.cs:
```diff
@@ -8,6 +8,9 @@
 
 namespace System.Security
 {
+#if NETCOREAPP // This file is compiled into both System.Private.CoreLib (NetCoreApp) and System.Security.Permission (netstandard2.0)
```
```suggestion
#if SYSTEM_PRIVATE_CORELIB
```

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Security/Permissions/PermissionState.cs:
```diff
@@ -3,6 +3,9 @@
 
 namespace System.Security.Permissions
 {
+#if NETCOREAPP // This file is compiled into both System.Private.CoreLib (NetCoreApp) and System.Security.Permission (netstandard2.0)
```
```suggestion
#if SYSTEM_PRIVATE_CORELIB
```

jeffhandley committed:
Use SYSTEM_PRIVATE_CORELIB constant instead of NETCOREAPP

jeffhandley committed:
Remove the code comment on the ifdef

jeffhandley commented:
All feedback thus far has been addressed and I've addressed all of the issues the checks surfaced that were related to this; I believe the remaining failing checks are failing for unrelated issues. Thank you for all of the feedback and coaching along the way, @jkotas!!

This is now ready for broader review. I'm hoping to get it merged into master by the end of the week with intent of getting this ported in for Preview 8.

jkotas reviewed (approved):
Thank you!

GrabYourPitchforks commented on src/libraries/Common/src/System/Diagnostics/NetFrameworkUtils.cs:
```diff
@@ -63,7 +63,7 @@ private static bool SafeWaitForMutex(Mutex mutexIn, ref Mutex mutexOut)
         {
             bool ret;
-            RuntimeHelpers.PrepareConstrainedRegions();
+
```
This .cs file is consumed by _System.Diagnostics.PerformanceCounter.csproj_ and _System.Diagnostics.EventLog.csproj_, both of which target netstandard + netframework. The call to `PrepareConstrainedRegions` should remain for those targets. Might be easiest just to pragma suppress this instance.

jkotas replied:
These projects consume these files under `$(TargetFramework.StartsWith('$(NetCoreAppCurrent)')) or $(TargetFramework.StartsWith('netcoreapp2.0'))` condition, so it should be fine to remove it.

GrabYourPitchforks commented on src/libraries/Microsoft.Win32.SystemEvents/src/Microsoft/Win32/SystemEvents.cs:
```diff
@@ -1114,7 +1114,6 @@ private static void Shutdown()
             }
         }
 
-        [PrePrepareMethod]
```
_Microsoft.Win32.SystemEvents.csproj_ includes netstandard + netframework as one of its targets. May need to keep this attribute.

jeffhandley replied:
The file is only consumed there when `$(TargetsWindows)` is true.  @jkotas, can that be true when targeting netstandard or netframework?

jeffhandley replied:
I just found references to `net461-Windows_NT`, so I believe the answer here is yes.

GrabYourPitchforks commented on src/libraries/System.Data.Odbc/src/Common/System/Data/Common/AdapterUtil.Odbc.cs:
```diff
@@ -656,7 +656,6 @@ internal static int StringLength(string inputString)
             return ((null != inputString) ? inputString.Length : 0);
         }
 
-        [ReliabilityContract(Consistency.WillNotCorruptState, Cer.MayFail)]
```
This project targets netstandard + netframework.

jeffhandley replied:
The file is in an ItemGroup that only targets NetCoreApp.

GrabYourPitchforks commented on src/libraries/System.Data.Odbc/src/Common/System/Data/ProviderBase/DbBuffer.cs:
```diff
@@ -19,7 +19,7 @@ protected DbBuffer(int initialSize) : base(IntPtr.Zero, true)
             if (0 < initialSize)
             {
                 _bufferLength = initialSize;
-                RuntimeHelpers.PrepareConstrainedRegions();
```
This project targets netstandard + netframework.
(Comment applies file-wide.)

jeffhandley replied:
The file is in an ItemGroup that only targets NetCoreApp.

GrabYourPitchforks commented on src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcCommand.cs:
```diff
@@ -675,7 +675,6 @@ private OdbcDataReader ExecuteReaderObject(CommandBehavior behavior, string meth
                 bool mustRelease = false;
                 CNativeBuffer parameterBuffer = _cmdWrapper._nativeParameterBuffer;
 
-                RuntimeHelpers.PrepareConstrainedRegions();
```
This project targets netstandard + netframework.
(This comment applies file-wide.)

jeffhandley replied:
The file is in an ItemGroup that only targets NetCoreApp.

GrabYourPitchforks commented on src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcConnectionHandle.cs:
```diff
@@ -57,9 +57,7 @@ private ODBC32.RetCode AutoCommitOff()
 
-            // Avoid runtime injected errors in the following block.
             // must call SQLSetConnectAttrW and set _handleState
-            RuntimeHelpers.PrepareConstrainedRegions();
```
This project targets netstandard + netframework.
(This comment applies file-wide.)

jeffhandley replied:
The file is in an ItemGroup that only targets NetCoreApp.

GrabYourPitchforks commented on src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcDataReader.cs:
```diff
@@ -2344,7 +2344,6 @@ internal int RetrieveKeyInfo(bool needkeyinfo, QualifiedTableName qualifiedTable
             bool mustRelease = false;
             Debug.Assert(buffer.Length >= 264, "Native buffer to small (_buffer.Length < 264)");
 
-            RuntimeHelpers.PrepareConstrainedRegions();
```
This project targets netstandard + netframework.
(This comment applies file-wide.)

jeffhandley replied:
The file is in an ItemGroup that only targets NetCoreApp.

GrabYourPitchforks commented on src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcHandle.cs:
```diff
@@ -21,8 +21,6 @@ protected OdbcHandle(ODBC32.SQL_HANDLE handleType, OdbcHandle parentHandle) : ba
             ODBC32.RetCode retcode = ODBC32.RetCode.SUCCESS;
 
-            // using ConstrainedRegions to make the native ODBC call and AddRef the parent
-            RuntimeHelpers.PrepareConstrainedRegions();
```
This project targets netstandard + netframework.
(This comment applies file-wide.)

jeffhandley replied:
The file is in an ItemGroup that only targets NetCoreApp.

GrabYourPitchforks commented on src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/EventLogInternal.cs:
```diff
@@ -1250,7 +1250,7 @@ private void VerifyAndCreateSource(string sourceName, string currentMachineName)
             if (!EventLog.SourceExists(sourceName, currentMachineName, true))
             {
                 Mutex mutex = null;
-                RuntimeHelpers.PrepareConstrainedRegions();
```
This project targets netstandard + netframework.
(This comment applies file-wide.)

jeffhandley replied:
The file is in an ItemGroup that only targets NetCoreApp.

GrabYourPitchforks commented on src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/Reader/UnsafeNativeMethods.cs:
```diff
@@ -459,7 +459,6 @@ internal enum EvtSeekFlags
         internal static extern bool EvtCancel(EventLogHandle handle);
 
         [DllImport(WEVTAPI)]
-        [ReliabilityContract(Consistency.WillNotCorruptState, Cer.Success)]
```
This project targets netstandard + netframework.
(This comment applies file-wide.)

jeffhandley replied:
The file is in an ItemGroup that only targets NetCoreApp.

GrabYourPitchforks commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapSessionOptions.cs:
```diff
@@ -1093,7 +1093,6 @@ private static bool AddLdapHandleRef(LdapConnection ldapConnection)
             bool success = false;
             if (ldapConnection != null && ldapConnection._ldapHandle != null && !ldapConnection._ldapHandle.IsInvalid)
             {
-                RuntimeHelpers.PrepareConstrainedRegions();
```
This project targets netstandard + netframework.
(This comment applies file-wide.)

jeffhandley replied:
The file is in an ItemGroup that does not target netstandard.

My understanding of the `_$(NetFrameworkCurrent)` target framework is that is used for NuGet packaging so that the package can target netframework, but the code from this project will not actually be compiled or run when targeting netframework, as the GAC will already have this assembly under the same identity.

With that rationale, this file is only compiled for NetCoreApp.

GrabYourPitchforks commented on src/libraries/System.Reflection.Metadata/src/System.Reflection.Metadata.csproj:
```diff
@@ -8,8 +8,10 @@
-    <!-- DesignTimeBuild requires all the TargetFramework Derived Properties to not be present in the first property group. -->
+  <!-- DesignTimeBuild requires all the TargetFramework Derived Properties to not be present in the first property group. -->
   <PropertyGroup>
+    <!-- Constrained Execution Regions apply to netstandard2.0 but not netstandard1.1 or net5+ -->
```
What about _net461_ or _$(NetFrameworkCurrent)_?

jeffhandley replied:
Good catch.  How does this look?

```xml
<!-- Constrained Execution Regions only apply to netstandard2.0 and net4* -->
<DefineConstants Condition="$(TargetFramework.StartsWith('netstandard2.0')) or $(TargetFramework.StartsWith('net4'))">$(DefineConstants);FEATURE_CER</DefineConstants>
```

GrabYourPitchforks replied:
LGTM!

GrabYourPitchforks commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Internal/MemoryBlocks/MemoryMappedFileBlock.cs:
```diff
@@ -18,7 +18,7 @@ private sealed class DisposableData : CriticalDisposableObject
             {
                 // Make sure the current thread isn't aborted in between acquiring the pointer and assigning the fields.
-#if !NETSTANDARD1_1
+#if FEATURE_CER
```
Nit: To conditionally remove a CER, I'd instead do this:

```cs
#if FEATURE_CER
    RuntimeHelpers.PrepareConstrainedRegions();
    try { /* intentionally left blank */ }
    finally
#endif
    {
        /* remainder of logic */
    }
```

Saves the try / finally info from being embedded into the resulting assembly file. It's just minor cleanup, though, and it might even make sense for us to do it as a separate PR. Might even find some other empty try blocks lurking in other .cs files. :)

jeffhandley replied:
Good suggestion.  I'll go ahead and do that here; there aren't many of them.

GrabYourPitchforks commented on src/libraries/System.Runtime/ref/System.Runtime.cs:
```diff
@@ -10185,32 +10201,32 @@ public partial class CryptographicException : System.SystemException
     {
         Demand = 2,
         Assert = 3,
-        [System.ObsoleteAttribute("Deny is obsolete and will be removed in a future release of the .NET Framework. See https://go.microsoft.com/fwlink/?LinkID=155570 for more information.")]
```
These attributes are still present in _SecurityAction.cs_. Should probably remove them from there as well.

GrabYourPitchforks commented on src/libraries/System.Security.Permissions/ref/System.Security.Permissions.cs:
```diff
@@ -1125,17 +1338,23 @@ public sealed partial class PrincipalPermission : System.Security.IPermission, S
-#if CAS_OBSOLETIONS
-        [System.ObsoleteAttribute("PrincipalPermissionAttribute is not honored by the runtime and must not be used.", true, DiagnosticId = "MSLIB0002", UrlFormat = "https://aka.ms/dotnet-warnings/{0}")]
+#if NET50_OBSOLETIONS
+        [System.ObsoleteAttribute("PrincipalPermissionAttribute is not honored by the runtime and must not be used..", true, DiagnosticId = "SYSLIB0002", UrlFormat = "https://aka.ms/dotnet-warnings/{0}")]
```
Double period at the end of this sentence.

jeffhandley replied:
Good catch!

GrabYourPitchforks commented on src/libraries/System.Security.Permissions/src/System/Security/SecurityManager.cs:
```diff
@@ -13,6 +13,9 @@ public static partial class SecurityManager
         [Obsolete]
         public static bool SecurityEnabled { get; set; }
         public static bool CurrentThreadRequiresSecurityContextCapture() { return false; }
+#if NET50_OBSOLETIONS
```
Recommend obsoleting the entire type. All methods return _default(T)_.

safern commented:
cc: @ericstj who in https://github.com/dotnet/runtime/issues/39401#issuecomment-659650339 raised the question of why not mark as obsolete in downlevel targets for OOB packages.

jeffhandley commented:
@safern
> cc: @ericstj who in [#39401 (comment)](https://github.com/dotnet/runtime/issues/39401#issuecomment-659650339) raised the question of why not mark APIs as obsolete in downlevel targets for OOB packages.

We wouldn't be able to use the new `DiagnosticId` or `UrlFormat` properties on the `Obsolete` attribute.

safern commented:
> We wouldn't be able to use the new DiagnosticId or UrlFormat properties on the Obsolete attribute.

Right, makes sense. Maybe we could add a workaround to include internal obsolete attribute that contains that overload/properties in the assemblies that target downlevel?

ericstj commented:
> We wouldn't be able to use the new DiagnosticId or UrlFormat properties on the Obsolete attribute.

Right, that's a design decision that was made.  Is that intentional?  Folks don't care about doing obseletion for packages which expose their API to NETStandard?  We don't consider it a loophole/problem that folks can use the latest packages, but target NETStandard and avoid the obsolete warnings?

jeffhandley commented:
> Right, makes sense. Maybe we could add a workaround to include internal obsolete attribute that contains that overload/properties in the assemblies that target downlevel?

Potentially, but at best those (or vanilla `[Obsolete]` attributes) would produce `CA0618` diagnostics as is seen today with the existing `[Obsolete]` attribute, yielding a different experience for suppressing these obsoletions when targeting downlevel than when targeting .NET 5+.  When targeting downlevel, you wouldn't get the ability to suppress all of CER/CAS in one shot without suppressing other obsoletions as well.

> Right, that's a design decision that was made. Is that intentional?

The loophole was not previously considered (that I know of).  It's great to call out; thank you!

> Folks don't care about doing obseletion for packages which expose their API to NETStandard?

Package authors targeting NETStandard, they can obsolete their APIs in the traditional way if they would like to.

> We don't consider it a loophole/problem that folks can use the latest packages, but target NETStandard and avoid the obsolete warnings?

If they want to utilize the loophole, indeed they would be able to.  But if we create a bifurcated user experience for these obsoletions, I don't think we'll get as much value out of the [better obsoletion](https://github.com/dotnet/designs/blob/master/accepted/2020/better-obsoletion/better-obsoletion.md) feature. I think this would introduce the scenario of:

- Using a library targeting .NET Standard
- Running on .NET Framework 4.7.1
- The feature works, but an obsoletion warning would still be raised

ericstj commented:
> Potentially, but at best those (or vanilla [Obsolete] attributes) would produce CA0618 diagnostics as is seen today with the existing [Obsolete] attribute, yielding a different experience for suppressing these obsoletions when targeting downlevel than when targeting .NET 5+.

We could permit an internally defined System.Obsolete or similar  attribute with the additional properties.  Old compilers could ignore it and new compilers could give it the new experience.

> Package authors targeting NETStandard, they can obsolete their APIs in the traditional way if they would like to.

I’m more concerned with the effectiveness of our own usage. Lots of folks will be using our packages and targeting < net5.0 for a while.  Maybe they stay targeting Netstandard2.0 for a long time and never see our obsoletion until we suddenly break them in net6 or net7.  Even when picking up the latest SDKs and packages.  Folks have a different expectation for our packages.
 
> Using a library targeting .NET Standard
Running on .NET Framework 4.7.1
The feature works, but an obsoletion warning would still be raised

I don’t think so. The attribute needs to be on types and members, not type forwards. If the types and members exist in package for netframework then they **will** be obsoleted. If the types and members are defined in netstandard then forwarded in netframework then a netframework app would never see the attributes nor warnings. A netstandard library would warn that library author appropriately.

jeffhandley committed:
Apply feedback related to CER obsoletions

jeffhandley committed:
Apply feedback related to CAS

jeffhandley commented:
> If the types and members are defined in netstandard then forwarded in netframework then a netframework app would never see the attributes nor warnings. A netstandard library would warn that library author appropriately.

Thanks -- I didn't know that. The suggestion makes more sense to me now.

This is intriguing and worth further consideration, but it was called out as a non-goal in the [better obsoletion](https://github.com/dotnet/designs/blob/master/accepted/2020/better-obsoletion/better-obsoletion.md#non-goals) doc. Would it be OK with you if we moved this PR forward but then I submit an issue for us to consider down-leveling the obsoletions to netstandard2.0 in a separate issue, @ericstj?

@GrabYourPitchforks - I've addressed all of your feedback -- thanks for checking so many details!

jkotas commented:
> Would it be OK with you if we moved this PR forward 

Yes, this should be done as follow up - if we decide that it is the right thing to do.

jeffhandley commented on src/libraries/System.Security.Permissions/src/System/Security/SecurityManager.cs:
```diff
@@ -6,6 +6,9 @@
 
 namespace System.Security
 {
+#if NET50_OBSOLETIONS
```
Oops -- I missed the ref assembly for this round of changes to CAS.  Addressing...

jeffhandley committed:
Update ref assembly to reflect SecurityManager obsoletion

ericstj commented:
Yeah I’m ok with that

jeffhandley committed:
Fix regression in System.IO.Pipes.NamedPipeServerStream.Windows.RunAsClient

jeffhandley commented:
@GrabYourPitchforks - I found a regression where we had [code calling into](https://github.com/dotnet/runtime/pull/39269/files#diff-6188922c3bee952b1f9add019fb48a34L224) `RuntimeHelpers.ExecuteCodeWithGuaranteedCleanup`, where I'd removed that call but removing that call changed the behavior of the code. I pushed a commit that replicated the behavior that method provided and it fixed the regression, but I'm now second-guessing whether or not we should obsolete that helper method.  I don't actually see any CER logic in it.

```csharp
        [Obsolete(Obsoletions.ConstrainedExecutionRegionMessage, DiagnosticId = Obsoletions.ConstrainedExecutionRegionDiagId, UrlFormat = Obsoletions.SharedUrlFormat)]
        public static void ExecuteCodeWithGuaranteedCleanup(TryCode code, CleanupCode backoutCode, object? userData)
        {
            if (code == null)
                throw new ArgumentNullException(nameof(code));
            if (backoutCode == null)
                throw new ArgumentNullException(nameof(backoutCode));
            bool exceptionThrown = true;
            try
            {
                code(userData);
                exceptionThrown = false;
            }
            finally
            {
                backoutCode(userData, exceptionThrown);
            }
        }
```

I'm considering removing the obsolete attribute from it and reverting the change where we removed the call to it.  What do you think?

GrabYourPitchforks commented:
`ExecuteCodeWithGuaranteedCleanup` call sites shouldn't be removed, since the code has a side effect. Instead, the call sites should be changed to use normal try / finally semantics instead of using this helper routine.

In .NET Framework, the implementation of this method made guarantees about how this code could execute within CERs. The implementation was, roughly:

```cs
PrepareDelegate(tryMethod);
PrepareDelegate(finallyMethod);

PrepareConstrainedRegions();
try
{
    tryMethod(userData);
}
finally
{
    finallyMethod(userData);
}
```

In .NET Core, since CER guarantees cannot be made, it's still a good idea to mark the method as obsolete in the sense of "this method doesn't do what you think it does."

jeffhandley commented:
The only check failing is because of the following, which is unrelated to these changes and has been failing on other PRs as well.

```
Assertion failed '!foundDiff' in 'OpenSsl:GetSslError(Microsoft.Win32.SafeHandles.SafeSslHandle,int,byref):int' during 'Linear scan register alloc' (IL size 99)
```

jeffhandley commented:
@ericstj I filed https://github.com/dotnet/designs/issues/145 for consider the downlevel obsoletion for OOB package APIs.

jeffhandley commented:
Well shoot -- it appears I somehow missed the "Classes that derive from CodeAccessSecurityAttribute" category.  I'll submit a follow-up PR for those.

jeffhandley commented:
https://github.com/dotnet/docs/issues/20894 submitted for the breaking change document.
